src/content/post/whatsapp-chatbot.mdx

```diff
@@ -0,0 +1,477 @@
+
+  const twiml = new MessagingResponse();
+  const aiReply = await reply(message.Body);
+
+  twiml.message(aiReply);
+  res.status(200).type('text/xml');
+  res.end(twiml.toString());
+});
+
+app.listen(3000, () => {
+  console.log('Express server listening on port 3000');
+});
+
+const chatMessages = [
+  {
+    role: 'system',
+    content: 'reply to the messages you get in 100 characters',
+  },
+];
+
+async function reply(msg) {
+  chatMessages.push({
+    role: 'user',
+    content: msg,
+  });
+  const response = await openai.chat.completions.create({
+    messages: chatMessages,
+    model: 'gpt-3.5-turbo',
+    max_tokens: 300,
+    temperature: 0.5,
+    frequency_penalty: 0.5,
+  });
+  return response.choices[0].message.content;
+}
+```
+
+Now, return to your terminal and execute the following command:
+
+```bash
+node server.js
+```
+
+If you have configured everything correctly, as we mentioned. You will see this message printed on the console.
+
+```
+Express server listening on port 3000
+```
+
+But there is a problem. Your web server is currently running locally on your computer. As a result, Twilio will not be able to access our webhook endpoint.
+
+This is where we need to use **Ngrok** to expose our local server to the internet. If you haven't used it before, [Twilio has already provided instructions on how to install and configure it on your machine](https://www.twilio.com/docs/usage/tutorials/how-to-set-up-your-node-js-and-express-development-environment). Once it's done (assuming the local server is still running), open another terminal in the same path and execute:
+
+```bash
+ngrok http 3000
+```
+
+This will expose our server to the internet. From the terminal, you can copy the forwarding URL with HTTPS and paste it into [Twilio WhatsApp Sandbox settings](https://console.twilio.com/us1/develop/sms/try-it-out/whatsapp-learn). Don't forget to add the `/incoming` route at the end and set the method to POST, then click Save.
+
+![Twilio whatsapp sandbox settings](~/assets/images/blog/whatsapp-chatbot/sandbox-settings.webp)
+
+Now, go back to the Twilio Whatsapp Sandbox tab and scan the QR code.
+
+![Twilio whatsapp sandbox](~/assets/images/blog/whatsapp-chatbot/sandbox.webp)
+
+From your WhatsApp, now you can chat with our AI just as you would normally with ChatGPT.
+
+![Twilio whatsapp chat demo v1](~/assets/images/blog/whatsapp-chatbot/chatbot-demo-v1.gif)
+
+## Adding Knowledge base
+
+A big challenge of working with embeddings is that traditional relational databases like MySQL or PostgreSQL cannot handle the complexity and scale of all that data. Therefore, AI engineers need a specialized storage system to efficiently handle high-dimensional vectors, and that's where vector databases come in.
+
+Now, let's create a table to store the vectors. Copy and paste the following code into your SQL editor (this editor allows you to write and run SQL queries and scripts on your database), and then execute it.
```

Comment:
I see the screen recording a little bit later that shows you writing some SQL in Supabase, but it would help developers to leave some links where they can access the same editor inside their Supabase project.